DllImportGenerator/DllImportGenerator/Marshalling/MarshallerHelpers.cs

```diff
@@ -86,10 +88,152 @@ public static TypeSyntax GetCompatibleGenericTypeParameterSyntax(this TypeSyntax
+            List<T> nodeIdToElement = new(elements.Count);
+            EdgeMap edgeMap = new EdgeMap(elements.Count);
+
+            int nextEdgeMapIndex = 0;
+            foreach (var element in elements)
+            {
+                elementIndexToEdgeMapNodeId.Add(keyFn(element), nextEdgeMapIndex++);
+                nodeIdToElement.Add(element);
+            }
+
+            foreach (var element in elements)
+            {
+                U elementIndex = keyFn(element);
+                foreach (var dependentElementIndex in getDependentIndicesFn(element))
+                {
+                    // Add an edge from the node for dependentElementIndex-> the node for elementIndex
+                    // This way, elements that have no dependencies have no edges pointing to them.
+                    edgeMap[elementIndexToEdgeMapNodeId[elementIndex], elementIndexToEdgeMapNodeId[dependentElementIndex]] = true;
+                }
+            }
+
+            // Now that we have initialized our map of edges and we have our list of nodes,
+            // we'll use Khan's algorithm to calculate a topological sort of the elements.
+            // Algorithm adapted from A. B. Kahn. 1962. Topological sorting of large networks. Commun. ACM 5, 11 (Nov. 1962), 558�562. DOI:https://doi.org/10.1145/368996.369025
+
+            // L is the sorted list
+            List<T> L = new List<T>(elements.Count);
+            // S is the set of elements with no incoming edges (no dependencies on it)
+            List<T> S = new List<T>(elements.Count);
+
+            // Initialize S
+            for (int node = 0; node < nodeIdToElement.Count; node++)
+            {
+                if (!edgeMap.AnyIncomingEdge(node))
+                {
+                    S.Add(nodeIdToElement[node]);
+                }
+            }
+
+            while (S.Count != 0)
+            {
+                // Remove element from S
+                T element = S[S.Count - 1];
+                S.RemoveAt(S.Count - 1);
+                // Add element to L
+                L.Add(element);
+                int n = elementIndexToEdgeMapNodeId[keyFn(element)];
+                // For each node m that n points to
+                for (int m = 0; m < edgeMap.NodeCount; m++)
+                {
+                    if (!edgeMap[m, n])
+                    {
+                        continue;
+                    }
+                    // Remove the edge from n to m
+                    edgeMap[m, n] = false;
+                    // If m does not have any incoming edges, add to S
+                    if (!edgeMap.AnyIncomingEdge(m))
+                    {
+                        S.Add(nodeIdToElement[m]);
+                    }
+                }
+            }
+
+            // If we have edges left, then we have a cycle.
+            if (edgeMap.AnyEdges)
+            {
+                throw new InvalidOperationException();
```

Comment:
Let's add a nice message here. This should be considered a catastrophic error for the source generator.